File: paddle/fluid/inference/anakin/convert/test_reshape_op.cc
/* Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License. */

#include <gtest/gtest.h>
#include "paddle/fluid/inference/anakin/convert/op_converter.h"
#include "paddle/fluid/inference/anakin/convert/ut_helper.h"

namespace paddle {
namespace inference {
namespace anakin {

template <typename TargetT>
void test_reshape1_op(const platform::DeviceContext& context, bool use_gpu) {
  framework::Scope scope;
  std::unordered_set<std::string> parameters;
  AnakinConvertValidation<TargetT, ::anakin::Precision::FP32> validator(
      parameters, &scope, context, use_gpu);

  // validator.DeclInputVar("reshape-X", {2, 3, 3, 1});
  // validator.DeclOutputVar("reshape-Out", {3, 2, 1, 3});
  validator.DeclInputVar("reshape-X", {1, 2, 4, 1});
  validator.DeclOutputVar("reshape-Out", {1, 8, 1, 1});

  framework::OpDesc desc;
  desc.SetType("reshape");
  desc.SetInput("X", {"reshape-X"});
  desc.SetOutput("Out", {"reshape-Out"});
  // desc.SetAttr("shape", std::vector<int>({3, 2, 1, 3}));
  desc.SetAttr("shape", std::vector<int>({1, 8, 1, 1}));

  LOG(INFO) << "set OP";
  validator.SetOp(*desc.Proto());
  LOG(INFO) << "execute";
  validator.Execute(1);
}

template <typename TargetT>
void test_reshape2_op(const platform::DeviceContext& context, bool use_gpu) {
  framework::Scope scope;
  std::unordered_set<std::string> parameters;
  AnakinConvertValidation<TargetT, ::anakin::Precision::FP32> validator(
      parameters, &scope, context, use_gpu);

  validator.DeclInputVar("reshape-X", {1, 2, 4});
  validator.DeclOutputVar("reshape-Out", {1, 4, 2});

  framework::OpDesc desc;
  desc.SetType("reshape");
  desc.SetInput("X", {"reshape-X"});
  desc.SetOutput("Out", {"reshape-Out"});
  // desc.SetAttr("shape", std::vector<int>({3, 2, 1, 3}));
  desc.SetAttr("shape", std::vector<int>({0, -1, 2}));

  LOG(INFO) << "set OP";
  validator.SetOp(*desc.Proto());
  LOG(INFO) << "execute";
  validator.Execute(1);
}

#ifdef PADDLE_WITH_CUDA
TEST(reshape1_op, gpu) {
  platform::CUDAPlace gpu_place(0);
  platform::CUDADeviceContext ctx(gpu_place);
  test_reshape1_op<::anakin::saber::NV>(ctx, true);
}

TEST(reshape2_op, gpu) {
  platform::CUDAPlace gpu_place(0);
  platform::CUDADeviceContext ctx(gpu_place);
  test_reshape2_op<::anakin::saber::NV>(ctx, true);
}
#endif

TEST(reshape1_op, cpu) {
  platform::CPUPlace cpu_place;
  platform::CPUDeviceContext ctx(cpu_place);
  test_reshape2_op<::anakin::saber::X86>(ctx, false);
}

TEST(reshape2_op, cpu) {
  platform::CPUPlace cpu_place;
  platform::CPUDeviceContext ctx(cpu_place);
  test_reshape2_op<::anakin::saber::X86>(ctx, false);
}

}  // namespace anakin
}  // namespace inference
}  // namespace paddle

USE_OP(reshape);
USE_CPU_ANAKIN_CONVERTER(reshape);

#ifdef PADDLE_WITH_CUDA
USE_ANAKIN_CONVERTER(reshape);
#endif

File: test/cinn/ops/test_reshape_op.py
#!/usr/bin/env python3

# Copyright (c) 2021 CINN Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from cinn.common import is_compiled_with_cuda
from cinn.frontend import NetBuilder
from op_test import OpTest, OpTestTool
from op_test_helper import TestCaseHelper

import paddle


@OpTestTool.skip_if(
    not is_compiled_with_cuda(), "x86 test will be skipped due to timeout."
)
class TestReshapeOp(OpTest):
    def setUp(self):
        print(f"\nRunning {self.__class__.__name__}: {self.case}")
        self.inputs = {}
        self.prepare_inputs()

    def prepare_inputs(self):
        self.inputs = {
            "x": self.random(self.case["shape"], self.case["dtype"]),
        }
        self.target_shape = self.case["target_shape"]

    def build_paddle_program(self, target):
        x = paddle.to_tensor(self.inputs["x"], stop_gradient=True)
        out = paddle.reshape(x, self.target_shape)
        self.paddle_outputs = [out]

    def build_cinn_program(self, target):
        builder = NetBuilder("reshape_test")
        x = builder.create_input(
            self.nptype2cinntype(self.inputs["x"].dtype),
            self.inputs["x"].shape,
            "x",
        )
        out = builder.reshape(x, self.target_shape)

        prog = builder.build()
        res = self.get_cinn_output(prog, target, [x], [self.inputs["x"]], [out])
        self.cinn_outputs = res

    def test_check_results(self):
        self.check_outputs_and_grads(all_equal=True)


class TestReshapeOpShape(TestCaseHelper):
    def init_attrs(self):
        self.class_name = "TestReshapeOpShape"
        self.cls = TestReshapeOp
        self.inputs = [
            # 1D -> [1-5]D
            {"shape": [100], "target_shape": [100]},
            {"shape": [100], "target_shape": [10, 10]},
            {"shape": [125], "target_shape": [5, 5, 5]},
            {"shape": [256], "target_shape": [4, 4, 4, 4]},
            {"shape": [1024], "target_shape": [8, 8, 4, 4]},
            # 2D -> [1-5]D
            {"shape": [5, 5], "target_shape": [25]},
            {"shape": [6, 8], "target_shape": [4, 12]},
            {"shape": [10, 20], "target_shape": [5, 10, 4]},
            {"shape": [4, 8], "target_shape": [2, 2, 2, 4]},
            {"shape": [16, 16], "target_shape": [4, 2, 2, 1, 16]},
            # 3D -> [1-5]D
            {"shape": [1, 1, 1], "target_shape": [1]},
            {"shape": [1, 2, 3], "target_shape": [6, 1]},
            {"shape": [4, 8, 16], "target_shape": [16, 8, 4]},
            {"shape": [6, 6, 6], "target_shape": [4, 9, 2, 3]},
            {"shape": [8, 1, 8], "target_shape": [2, 2, 2, 2, 4]},
            # 4D -> [1-5]D
            {"shape": [4, 1, 2, 1], "target_shape": [8]},
            {"shape": [2, 2, 4, 8], "target_shape": [4, 32]},
            {"shape": [6, 7, 8, 9], "target_shape": [42, 36, 2]},
            {"shape": [1024, 1, 1, 1], "target_shape": [4, 4, 8, 8]},
            {"shape": [10, 20, 30, 40], "target_shape": [8, 6, 4, 2, 625]},
            # special
            {"shape": [1, 1024, 4], "target_shape": [1, 2048, 2]},
            {"shape": [2048, 2, 2], "target_shape": [256, 8, 4]},
            {"shape": [1, 1, 256], "target_shape": [16, 1, 16]},
            {"shape": [1, 1, 1, 1], "target_shape": [1, 1]},
            {"shape": [1, 1, 1], "target_shape": [1]},
            {"shape": [1], "target_shape": [1, 1, 1, 1]},
        ]
        self.dtypes = [
            {"dtype": "float32"},
        ]
        self.attrs = []


class TestReshapeOpDtype(TestCaseHelper):
    def init_attrs(self):
        self.class_name = "TestReshapeOpDtype"
        self.cls = TestReshapeOp
        self.inputs = [
            {"shape": [2, 3, 4], "target_shape": [4, 6]},
        ]
        self.dtypes = [
            {"dtype": "float16"},
            {"dtype": "float32"},
            {"dtype": "float64"},
            {"dtype": "bool"},
            {"dtype": "uint8"},
            {"dtype": "int8"},
            {"dtype": "int32"},
            {"dtype": "int64"},
        ]
        self.attrs = []


if __name__ == "__main__":
    TestReshapeOpShape().run()
    TestReshapeOpDtype().run()
